__tests__/setup-elixir.test.js

```diff
@@ -1,21 +1,178 @@
-const {getElixirVersion} = require('../src/setup-elixir')
-const {deepStrictEqual} = require('assert')
+simulateInput('otp-version', '23.2')
+simulateInput('elixir-version', '1.11')
+simulateInput('rebar3-version', '3.14')
+simulateInput('install-rebar', 'true')
+simulateInput('install-hex', 'true')
 
-async function test() {
-  let vsn
-  vsn = await getElixirVersion('v1.10.x', '23')
-  deepStrictEqual(vsn, ['v1.10.4', '23'])
+const setupElixir = require('../src/setup-elixir')
+const assert = require('assert')
+const installer = require('../src/installer')
+const path = require('path')
 
-  vsn = await getElixirVersion('^v1.10', '23')
-  deepStrictEqual(vsn, ['v1.10.4', '23'])
+async function all() {
+  await testFailInstallOTP()
+  await testFailInstallElixir()
+  await testFailInstallRebar3()
 
-  vsn = await getElixirVersion('v1.11.0-rc.0', '23')
-  deepStrictEqual(vsn, ['v1.11.0-rc.0', '23'])
+  await testOTPVersions()
+  await testElixirVersions()
+  await testRebar3Versions()
 }
 
-test()
+async function testFailInstallOTP() {
+  const otpOSVersion = 'ubuntu-08.04'
+  const otpVersion = 'OTP-23.2'
+  await assert.rejects(
+    async () => {
+      await installer.installOTP(otpOSVersion, otpVersion)
+    },
+    (err) => {
+      assert.ok(err instanceof Error)
+      return true
+    },
+    `Installing Erlang/OTP ${otpVersion} over ${otpOSVersion} is supposed to fail`,
+  )
+}
+
+async function testFailInstallElixir() {
+  let exVersion
+
+  exVersion = '0.11'
+  await assert.rejects(
+    async () => {
+      await installer.installElixir(exVersion)
+    },
+    (err) => {
+      assert.ok(err instanceof Error)
+      return true
+    },
+    `Installing Elixir ${exVersion} is supposed to fail`,
+  )
+
+  exVersion = 'v1.0.0-otp-17'
+  await assert.rejects(
+    async () => {
+      await installer.installElixir(exVersion)
+    },
+    (err) => {
+      assert.ok(err instanceof Error)
+      return true
+    },
+    `Installing Elixir ${exVersion} is supposed to fail`,
+  )
+}
+
+async function testFailInstallRebar3() {
+  const r3Version = '0.14.4'
+  await assert.rejects(
+    async () => {
+      await installer.installRebar3(r3Version)
+    },
+    (err) => {
+      assert.ok(err instanceof Error)
+      return true
+    },
+    `Installing rebar3 ${r3Version} is supposed to fail`,
+  )
+}
+
+async function testOTPVersions() {
+  const otpBuildsListing = listing('otp_builds.txt')
```

Comment:
Since we require network connection for other tests what about dropping these files and fetching the actual files for simplicity. That way we will also catch any issues with changes to the files.

Reply:
We can do that, sure. I'm just afraid that changing some of these will actually lead to different test results (e.g. I can't use `1` as a version since that'll yield different results as soon as `builds.txt` changes), but I can do it, no prob.

Reply:
f79efb2: done. I'm afraid tests lost some quality, though, as per what I explained previously (if depending on a dynamic environment we can't actually be sure of what'll happen when we do stuff like `'^1.10'`)

Reply:
We can do the tests on version ranges that are no longer supported and won't be updated.

Reply:
That's what supposedly did, with this change. In any case, to be completely sure, I'd have to know exactly what isn't supported. We can deal with mishaps, if any, later, if new versions come out that change test results.

Reply:
You can find supported Elixir versions here: https://hexdocs.pm/elixir/compatibility-and-deprecations.html#content. I would guess Erlang has similar documentation.

Reply:
OT: when you resolve conservations after responding it is very hard to find your replies.

Reply:
> OT: when you resolve conservations after responding it is very hard to find your replies.

Sorry about that. It sometimes happens. Is anything missing regarding this conversation, in particular?

Reply:
This is Erlang/OTP testing: https://github.com/erlef/setup-beam/pull/9/files#diff-c76af13c0a24fd165fe6700b8b1254b6a48a60b9ddfcb39dcad476a725f8106bR84 (all results belong to no-longer-supported versions)
This is Elixir testing: https://github.com/erlef/setup-beam/pull/9/files#diff-c76af13c0a24fd165fe6700b8b1254b6a48a60b9ddfcb39dcad476a725f8106bR109 (all results belong to no-longer-supported versions)
This is `rebar3` testing: https://github.com/erlef/setup-beam/pull/9/files#diff-c76af13c0a24fd165fe6700b8b1254b6a48a60b9ddfcb39dcad476a725f8106bR133 (all results belong to no-longer-supported versions)

I assume we're good (?)

Reply:
All good!